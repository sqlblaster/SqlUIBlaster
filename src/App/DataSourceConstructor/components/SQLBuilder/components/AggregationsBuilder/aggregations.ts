export enum AggregationOperatorEnum {
  'Count of rows',
  'Sum of',
  'Average of',
  'Number of distinct values of',
  'Cumulative sum of',
  'Cumulative count of rows',
  'Standard deviation of',
  'Minimum of',
  'Maximum of'
}

export type AggregationOperator = keyof typeof AggregationOperatorEnum;

export const aggregationOperators: AggregationOperator[] = [
  'Count of rows',
  'Sum of',
  'Average of',
  'Number of distinct values of',
  'Cumulative sum of',
  'Cumulative count of rows',
  'Standard deviation of',
  'Minimum of',
  'Maximum of'
];
