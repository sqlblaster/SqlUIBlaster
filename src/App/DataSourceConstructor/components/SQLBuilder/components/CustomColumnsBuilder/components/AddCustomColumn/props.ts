import { actionCreators, PickActionCreators } from 'src/components/query-builder/store/action-creators';
import { PickStates } from 'src/components/query-builder/store/models/State';
import { PopoverManagerProps } from '../../../common/WithPopoverManagement';

export type AddCustomColumnOwnProps = PopoverManagerProps;

export type AddCustomColumnStateProps = PickStates<'customColumns'>;

export type AddCustomColumnDispatchProps = PickActionCreators<'addCustomColumn'>;

export const AGMapDispatchToProps: AddCustomColumnDispatchProps = {
  addCustomColumn: actionCreators.addCustomColumn
};

export type AddCustomColumnProps = AddCustomColumnStateProps & AddCustomColumnDispatchProps;

export type AddCustomColumnViewProps = AddCustomColumnOwnProps &
  AddCustomColumnStateProps &
  AddCustomColumnDispatchProps;
