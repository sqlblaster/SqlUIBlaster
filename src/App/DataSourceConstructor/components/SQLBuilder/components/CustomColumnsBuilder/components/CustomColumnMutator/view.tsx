import * as React from 'react';
import { ColumnSelector } from '../../../common/ColumnSelector';
import { PopoverPaper } from '../../../common/PopoverPaper';
import { CustomColumnMutatorViewProps } from './props';

export const CustomColumnMutator: React.FC<
  CustomColumnMutatorViewProps
> = props => {
  const {
    popoverPositionUpdater,
    onCustomColumnSelected,
    canShowColumnBranch
  } = props;

  return (
    <PopoverPaper {...props}>
      <div className='mutate-customColumn'>
        <ColumnSelector
          popoverPositionUpdater={popoverPositionUpdater}
          onColumnSelected={onCustomColumnSelected}
          canShowColumnBranch={canShowColumnBranch}
        />
      </div>
    </PopoverPaper>
  );
};
