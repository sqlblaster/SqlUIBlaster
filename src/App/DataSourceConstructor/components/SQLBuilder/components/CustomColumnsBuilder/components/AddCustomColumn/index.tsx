import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/components/query-builder/store/models/State';
import { WithPopoverManagement } from '../../../common/WithPopoverManagement';
import {
  AddCustomColumnDispatchProps,
  AddCustomColumnProps,
  AddCustomColumnStateProps,
  AGMapDispatchToProps
} from './props';
import { AddCustomColumn as View } from './view';

export const AddCustomColumn = connect<
  AddCustomColumnStateProps,
  AddCustomColumnDispatchProps,
  {},
  State
>(
  ({ customColumns }) => ({
    customColumns
  }),
  AGMapDispatchToProps
)(
  class extends React.Component<AddCustomColumnProps> {
    public static ViewWithPopoverManagement = WithPopoverManagement(View);

    public render() {
      return (
        <AddCustomColumn.ViewWithPopoverManagement
          {...this.props}
        />
      );
    }
  }
);
