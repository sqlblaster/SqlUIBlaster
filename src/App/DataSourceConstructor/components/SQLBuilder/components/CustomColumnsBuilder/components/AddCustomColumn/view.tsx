import * as React from 'react';
import { AddEntity } from '../../../common/AddEntity';
import { CustomColumnMutator } from '../CustomColumnMutator';
import { AddCustomColumnViewProps } from './props';

export const addCustomColumnPlaceholder = 'Add a column';

// FIXME: make padding clickable too
export class AddCustomColumn extends React.Component<AddCustomColumnViewProps> {
  private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

  constructor(props: AddCustomColumnViewProps) {
    super(props);

    this.popoverLauncher = React.createRef<HTMLDivElement>();
  }

  public render() {
    const {
      customColumns,
      addCustomColumn,
      isOpen,
      handleClose,
      handleOpen
    } = this.props;

    return (
      <>
        <AddEntity
          showText={!(customColumns && customColumns.length)}
          onClick={handleOpen}
          reference={this.popoverLauncher}
        >
          {addCustomColumnPlaceholder}
        </AddEntity>
        { isOpen && (// TODO: resolve this problem another way
          <CustomColumnMutator
            action={addCustomColumn}
            isOpen={isOpen}
            anchorEl={this.popoverLauncher.current}
            onClose={handleClose}
          />
        ) }
      </>
    );
  }
}
