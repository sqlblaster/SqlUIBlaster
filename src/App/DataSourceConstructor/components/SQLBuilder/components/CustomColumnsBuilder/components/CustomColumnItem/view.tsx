import { withStyles } from '@material-ui/core';
import * as React from 'react';
import { RemoveButton } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/Icons/RemoveButton';
import { SequentialColumnView } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/SequentialColumnView';
import { CustomColumnMutator } from '../CustomColumnMutator';
import { CustomColumnItemViewProps } from './props';
import { customColumnItemStyles } from './styles';

export const CustomColumnItem = withStyles(customColumnItemStyles)(
  class extends React.Component<CustomColumnItemViewProps> {
    private readonly popoverLauncher: React.RefObject<HTMLDivElement>;

    constructor(props: CustomColumnItemViewProps) {
      super(props);

      this.popoverLauncher = React.createRef<HTMLDivElement>();
    }

    public render() {
      const {
        customColumn,
        updateCustomColumn,
        classes,
        isOpen,
        handleOpen,
        handleClose,
        handleCustomColumnRemoval
      } = this.props;

      const { id, column } = customColumn;

      return (
        <>
          <div className={classes.root}>
            <div
              ref={this.popoverLauncher}
              onClick={handleOpen}
              className={classes['customColumn-view']}
              data-testid='customColumn-view'
            >
              <SequentialColumnView {...column} />
            </div>
            <RemoveButton
              onClick={handleCustomColumnRemoval(id)}
              data-testid='remove'
            />
          </div>
          {isOpen && (
            <CustomColumnMutator
              isOpen={isOpen}
              anchorEl={this.popoverLauncher.current}
              action={updateCustomColumn}
              onClose={handleClose}
              customColumn={customColumn}
            />
          )}
        </>
      );
    }
  }
);
