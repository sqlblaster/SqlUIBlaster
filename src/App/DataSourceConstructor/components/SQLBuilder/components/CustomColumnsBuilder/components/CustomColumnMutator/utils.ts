import { ColumnBranch } from '../../../common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { CustomColumn } from '../CustomColumnItem/model';

export const filterCustomColumn = (
  customColumnBranch: ColumnBranch,
  customColumns: CustomColumn[]
): boolean =>
  !customColumns.find(({ column }) => column.equals(customColumnBranch)) &&
  customColumnBranch !== customColumnBranch.lastColumn;
