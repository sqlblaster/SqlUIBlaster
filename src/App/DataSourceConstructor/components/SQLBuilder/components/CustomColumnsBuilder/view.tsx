import { Typography, withStyles } from '@material-ui/core';
import * as React from 'react';
import { AddCustomColumn } from './components/AddCustomColumn';
import { CustomColumnItem } from './components/CustomColumnItem';
import { CustomColumnsBuilderViewProps } from './props';
import { customColumnsBuilderStyles } from './styles';

export const hiddenCustomColumnsText = `Can't add columns while there are aggregations or groupings`;

export const CustomColumnsBuilder = withStyles(customColumnsBuilderStyles)((({
  customColumns,
  aggregations,
  groupings,
  classes
}) => (
  <>
    {!(
      (aggregations && aggregations.length) ||
      (groupings && groupings.length)
    ) ? (
      <div className={classes.root}>
        {customColumns && customColumns.length ? (
          <div className={classes.customColumns}>
            {customColumns.map(customColumn => (
              <CustomColumnItem
                key={customColumn.id}
                customColumn={customColumn}
              />
            ))}
          </div>
        ) : null}
        <AddCustomColumn />
      </div>
    ) : (
      <Typography className={classes['no-column']}>
        {hiddenCustomColumnsText}
      </Typography>
    )}
  </>
)) as React.FC<CustomColumnsBuilderViewProps>);
