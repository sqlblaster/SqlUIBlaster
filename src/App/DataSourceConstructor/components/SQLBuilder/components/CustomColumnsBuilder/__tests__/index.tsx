import {
  fireEvent,
  getByText as getByTextOn,
  queryByText as queryByTextOn
} from 'react-testing-library';
import { regex } from 'src/App/DataSourceConstructor/test-utils/utils';
import { getGeneratedQuery, render } from '../../../__tests__/render.utils';
import { customColumnsBuilderLabel } from '../../../labels';
import { aggregationOperators } from '../../AggregationsBuilder/aggregations';
import {
  clickAddAAggregation,
  getAggregationsBuilderPanel
} from '../../AggregationsBuilder/components/AddAggregation/__tests__/utils/click-add-a-aggregation';
import { selectAColumn } from '../../common/ColumnSelector/__tests__/utils/select-a-column';
import {
  clickAddAGrouping,
  getGroupingsBuilderPanel
} from '../../GroupingsBuilder/components/AddGrouping/__tests__/utils/click-add-a-grouping';
import { selectATable } from '../../TableSelector/__tests__/utils/select-table';
import {
  clickAddACustomColumn,
  clickAddMoreCustomColumn
} from '../components/AddCustomColumn/__tests__/utils/click-add-a-custom-column';
import { addCustomColumnPlaceholder } from '../components/AddCustomColumn/view';
import { hiddenCustomColumnsText } from '../view';

describe('Custom Columns Builder', () => {
  it(`should be able to
        shows the add custom column placeholder,
        add the 1 lvl custom column,
        hide the add custom column placeholder,
        update custom column to another one,
        add one more 1 lvl custom column,
        check that 1st 1 lvl column is not available,
        remove that custom column,
        hide custom columns on aggregations or groupings presence,
        show previous added custom columns
        after all aggregations and groupings removed`, async () => {
    const { getByText, queryByText, getAllByText } = render();

    const columns = selectATable();

    const customColumnsBuilderPanel = getByTextOn(
      document.body,
      regex(customColumnsBuilderLabel)
    );

    const clickOnCustomColumn = (columnName: string) => {
      fireEvent.click(
        getByTextOn(customColumnsBuilderPanel, regex(columnName))
      );
    };

    // shows the add custom column placeholder
    expect(getByText(addCustomColumnPlaceholder)).toBeInTheDocument();

    // add the 1 lvl custom column
    clickAddACustomColumn();
    const { foreignColumns, foreignColumnsPanel } = selectAColumn(columns[5], {
      expand: true,
      container: customColumnsBuilderPanel
    });
    selectAColumn(foreignColumns[1], {
      container: foreignColumnsPanel
    });

    expect(getGeneratedQuery()).toMatchSnapshot('add the 1 lvl custom column');

    // hides the add custom column placeholder
    expect(queryByText(addCustomColumnPlaceholder)).not.toBeInTheDocument();

    // update custom column to another one
    clickOnCustomColumn(foreignColumns[1].name);
    const { foreignColumnsPanel: foreignColumnsPanel1 } = selectAColumn(
      columns[5],
      {
        expand: true,
        container: customColumnsBuilderPanel
      }
    );
    selectAColumn(foreignColumns[2], {
      container: foreignColumnsPanel1
    });

    expect(getGeneratedQuery()).toMatchSnapshot(
      'update custom column to another one'
    );

    // add one more 1 lvl custom column
    clickAddMoreCustomColumn();

    // check that 1st 1 lvl column is not available
    const { foreignColumnsPanel: foreignColumnsPanel2 } = selectAColumn(
      columns[5],
      { expand: true }
    );
    if (!foreignColumnsPanel2) {
      throw new Error(`Couldn't get Foreign column fields panel`);
    }
    expect(
      queryByTextOn(foreignColumnsPanel2, regex(foreignColumns[2].name))
    ).toBeFalsy();

    const {
      foreignColumns: foreignColumns2,
      foreignColumnsPanel: foreignColumnsPanel3
    } = selectAColumn(columns[6], {
      expand: true
    });
    selectAColumn(foreignColumns2[2], { container: foreignColumnsPanel3 });

    expect(getGeneratedQuery()).toMatchSnapshot(
      'add one more 1 lvl custom column'
    );

    // remove custom column
    fireEvent.click(getAllByText(/clear/i)[0]);

    expect(getGeneratedQuery()).toMatchSnapshot('remove custom column');

    // hide custom columns on aggregations or groupings presence
    clickAddAAggregation();
    fireEvent.click(getByText(regex(aggregationOperators[5])));

    expect(customColumnsBuilderPanel).toHaveTextContent(
      hiddenCustomColumnsText
    );

    clickAddAGrouping();
    selectAColumn(columns[3]);

    expect(customColumnsBuilderPanel).toHaveTextContent(
      hiddenCustomColumnsText
    );

    // show previous added custom columns after all aggregations and groupings removed
    fireEvent.click(getByTextOn(getAggregationsBuilderPanel(), /clear/i));

    expect(customColumnsBuilderPanel).toHaveTextContent(
      hiddenCustomColumnsText
    );

    fireEvent.click(getByTextOn(getGroupingsBuilderPanel(), /clear/i));

    expect(customColumnsBuilderPanel).not.toHaveTextContent(
      hiddenCustomColumnsText
    );
  });
});
