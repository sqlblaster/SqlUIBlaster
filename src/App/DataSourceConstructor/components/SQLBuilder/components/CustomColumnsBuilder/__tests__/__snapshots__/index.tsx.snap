// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Columns Builder should be able to
        shows the add custom column placeholder,
        add the 1 lvl custom column,
        hide the add custom column placeholder,
        update custom column to another one,
        add one more 1 lvl custom column,
        check that 1st 1 lvl column is not available,
        remove that custom column,
        hide custom columns on aggregations or groupings presence,
        show previous added custom columns
        after all aggregations and groupings removed: add one more 1 lvl custom column 1`] = `
"SELECT
  orders1.*,
  \\"products1\\".\\"Available amount\\" AS \\"Available amount (products1)\\",
  \\"users1\\".\\"Username\\" AS \\"Username (users1)\\"
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\" ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
  LEFT JOIN \\"users\\" \\"users1\\" ON \\"orders1\\".\\"UserID\\" = \\"users1\\".\\"ID\\"
LIMIT 2000
"
`;

exports[`Custom Columns Builder should be able to
        shows the add custom column placeholder,
        add the 1 lvl custom column,
        hide the add custom column placeholder,
        update custom column to another one,
        add one more 1 lvl custom column,
        check that 1st 1 lvl column is not available,
        remove that custom column,
        hide custom columns on aggregations or groupings presence,
        show previous added custom columns
        after all aggregations and groupings removed: add the 1 lvl custom column 1`] = `
"SELECT
  orders1.*,
  \\"products1\\".\\"Product name\\" AS \\"Product name (products1)\\"
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\" ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
LIMIT 2000
"
`;

exports[`Custom Columns Builder should be able to
        shows the add custom column placeholder,
        add the 1 lvl custom column,
        hide the add custom column placeholder,
        update custom column to another one,
        add one more 1 lvl custom column,
        check that 1st 1 lvl column is not available,
        remove that custom column,
        hide custom columns on aggregations or groupings presence,
        show previous added custom columns
        after all aggregations and groupings removed: remove custom column 1`] = `
"SELECT
  orders1.*,
  \\"users1\\".\\"Username\\" AS \\"Username (users1)\\"
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"users\\" \\"users1\\" ON \\"orders1\\".\\"UserID\\" = \\"users1\\".\\"ID\\"
LIMIT 2000
"
`;

exports[`Custom Columns Builder should be able to
        shows the add custom column placeholder,
        add the 1 lvl custom column,
        hide the add custom column placeholder,
        update custom column to another one,
        add one more 1 lvl custom column,
        check that 1st 1 lvl column is not available,
        remove that custom column,
        hide custom columns on aggregations or groupings presence,
        show previous added custom columns
        after all aggregations and groupings removed: update custom column to another one 1`] = `
"SELECT
  orders1.*,
  \\"products1\\".\\"Available amount\\" AS \\"Available amount (products1)\\"
FROM
  \\"orders\\" \\"orders1\\"
  LEFT JOIN \\"products\\" \\"products1\\" ON \\"orders1\\".\\"ProductID\\" = \\"products1\\".\\"ID\\"
LIMIT 2000
"
`;
